### PR TITLE
Set up deployment pipeline with dev stage

### PR DESCRIPTION
**Pull request by beekley**
## Description

Currently, we have sandbox environments for each developer. We want to have environments that aren't specific to devs and have (relatively) current code.

This PR sets up a pipeline to deploy a `DEV` environment, which is just like the `SANDBOX` envs, but pulls from the `main` branch of this repo.

### New

- A new CDK app, `pipeline`.
- Added an `envType` variable to the stack, so that each environment has its own namespace.

### Changed

- Lowered the required Node version for the client to 16, since there was a dependency that needed <=17, which is [not supported anymore](https://nodejs.org/en/about/releases/).

## Testing and Reviewing

I could synth this CDK config, but not deploy it as I don't have write access to the Deployments account. The Github connector may need to be given access to this repo as well.

### CODE REVIEW

File: cdk/src/util.ts [hunk 1]
```diff
@@ -1,25 +1,35 @@
...
+}
 
 // Each construct in AWS is prefixed with its stack's name. This function creates a useful name
 // for identification of the construct.
-// TODO: drop developer user name from the stack for non-dev environments.
-export const stackName = (id: StackId): string => {
+export const stackName = (id: StackId, e: EnvType): string => {
   const base = `${projectName}${StackId[id]}`;
-  // Share a common Network name across multiple developers, so don't include the user name.
+  // Network stuff should be shared across environments within a given AWS account, so use a
```
Comment (kailamjeter) [inline cdk/src/util.ts:30]: also between environments?

**Review by kailamjeter — APPROVED** [review]

File: cdk/src/pipeline/pipeline-stack.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+          // Update npm.
+          'npm i -g npm',
+
+          // Build the client.
+          'npm --prefix client ci',
+          'npm --prefix client run build',
+
+          // Synthesize the CDK app.
+          'npm --prefix run cdk ci',
+          'npm --prefix run cdk pipeline-synth',
```
Comment (arockett) [inline cdk/src/pipeline/pipeline-stack.ts:34]: I haven't used the ```--prefix``` option before, but wouldn't this need to be

```
'npm --prefix cdk ci',
'npm --prefix cdk run pipeline-synth'
```

File: cdk/src/pipeline/pipeline-stack.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        ],
+        primaryOutputDirectory: 'cdk/cdk.out',
+      }),
+    });
+
+    pipeline.addStage(
+      new OpenDataPlatformStage(scope, 'Development', {
```
Comment (arockett) [inline cdk/src/pipeline/pipeline-stack.ts:41]: Please us the id 'Dev' so we can keep the stack names short and consistent across projects

File: cdk/src/open-data-platform/root-stack.ts [hunk 4]
```diff
@@ -9,22 +9,27 @@ import { FrontendStack } from './frontend/frontend-stack';
 import { AppPlaneStack } from './app-plane/app-plane-stack';
 
 export class OpenDataPlatformStack extends Stack {
-  constructor(scope: Construct, id: string, props?: util.CommonProps) {
+  constructor(scope: Construct, id: string, props: util.CommonProps) {
     super(scope, id, props);
+    const { envType } = props;
 
-    const networkStack = new NetworkStack(scope, util.stackName(util.StackId.Network), {
+    const networkStack = new NetworkStack(scope, util.stackName(util.StackId.Network, envType), {
```
Comment (arockett) [inline cdk/src/open-data-platform/root-stack.ts:16]: Is 'scope' instead of 'this' used intentionally here and in other places? What is the difference?

**Review by arockett — CHANGES_REQUESTED** [review]
Mostly looks good but I'm wondering how the stack naming utilities plus pipeline stage and nested stack structure you've gone with will effect the end stack names which are deployed through the pipeline.

Since this cdk project's stage and stacks aren't set up exactly the same as the Patina/Tributary pipelines, there may be a bit of a back and forth up front to make sure the pipeline deploys with the names we want. You and I should schedule some time to just knock out that iteration for early next week.

File: cdk/src/open-data-platform/root-stack.ts (hunk 4 above)
Comment (beekley) [reply cdk/src/open-data-platform/root-stack.ts:16]: (as discussed) this was unintentional and meant that each of these stacks aren't actually nested stacks.

I've removed the root stack altogether now.

File: cdk/src/pipeline/pipeline-stack.ts (hunk 2 above)
Comment (beekley) [reply cdk/src/pipeline/pipeline-stack.ts:34]: Good catch, thanks.

File: cdk/src/util.ts (hunk 1 above)
Comment (beekley) [reply cdk/src/util.ts:28]: I've rephrased.

File: cdk/src/util.ts [hunk 5]
```diff
@@ -1,25 +1,35 @@
...
+  // common name for the network stack.
   if (id === StackId.Network) return base;
-  return `${process.env.USER}-${base}`;
+  if (e == EnvType.Sandbox && process.env.USER) return `${process.env.USER}-${base}`;
+  return `${e}-${base}`;
```
[the remark this answers is not in the document]
Comment (beekley) [reply cdk/src/util.ts]: Thanks, I've removed that prefix.

**Review by arockett — APPROVED** [review]
LGTM, let's schedule time to init the pipeline in the Deployments account and see if everything works as expected

**Pull request merged**